### PR TITLE
Add actions support to package auth verification

### LINKED ISSUE #23642: actions: login via GITEA_TOKEN to package repository not working
### Description

I am having this snippet in a workflow file:
```yaml
      - name: Log in to the Container registry
        uses: docker/login-action@v2
        with:
          registry: my.gitea.domain
          username: ${{ github.repository_owner }}
          password: ${{ secrets.GITEA_TOKEN }}
 ```
I exchanged `github.repository_owner` with `github.actor` as well. Bot did not work.

I see the following error in the actions run log:
```
Log in to the Container registry
Logging into my.gitea.domain...
::error::Error response from daemon: login attempt to https://my.gitea.domain/v2/ failed with status: 401 Unauthorized
```

### Gitea Version

1.19.0

### Can you reproduce the bug on the Gitea demo site?

No

### Log Gist

_No response_

### Screenshots

_No response_

### Git Version

_No response_

### Operating System

_No response_

### How are you running Gitea?

I am running it in a docker container and am reverse proxying the https port.

### Database

None

### PR DESCRIPTION
**Pull request by yp05327**
Partly fixes https://github.com/go-gitea/gitea/issues/23642

Error info:
image
ActionsUser (userID -2) is used to login in to docker in action jobs.

Due to we have no permission policy settings of ActionsUser now, ActionsUser can only access public registry by this quick fix.

### CODE REVIEW

**Review by lunny — DISMISSED** [review]

File: routers/api/packages/container/auth.go [hunk 1]
```diff
@@ -27,11 +27,16 @@ func (a *Auth) Verify(req *http.Request, w http.ResponseWriter, store auth.DataS
...
-	if uid == 0 {
-		return nil, nil
-	}
-	if uid == -1 {
-		return user_model.NewGhostUser(), nil
+	if uid <= 0 {
+		switch uid {
+		case -1:
+			return user_model.NewGhostUser(), nil
+		case -2:
```
Comment (wxiaoguang) [inline routers/api/packages/container/auth.go]:
```suggestion
		case user_model.ActionsUserID:
```

?

IIRC there is a const for it.

Comment (yp05327) [reply routers/api/packages/container/auth.go]: I have a question, why `GhostUser` is not defined by consts.

**Comment by yp05327** [comment]
I found that we already have a func called `GetPossibleUserByID` to handle with both normal users and system users.

**Review by wolfogre — APPROVED** [review]

**Review by KN4CK3R — CHANGES_REQUESTED** [review]
This will only work if you pull images from a public registry. And you can do that without auth. The action user will not pass permission checks.

**Comment by wxiaoguang** [comment]
> This will only work if you pull images from a public registry. And you can do that without auth. The action user will not pass permission checks.

That's also an unclear question in my mind, how the "action" user works with permissions (or how it should work).

Is there any document?

**Comment by KN4CK3R** [comment]
open questions I asked in Discord:

> the first question is if that task token should/could be used to access the package registry
if the answer is yes, we need a definition of who is the real user behind that task
is it the owner of the repo, the user which triggered the task?

And after that we may need to change how we use `ctx.Doer`. It's not clear to me where the actions user should be the doer or the real user.

**Comment by yp05327** [comment]
So `GITEA_TOKEN` needs to be `run.TriggerUser`'s temporary token instead of `task.Token`?

**Review by yardenshoham — APPROVED** [review]

**Comment by yp05327** [comment]
`Team.Authorize` and `TeamUnit.AccessMode` confused me.
It seems that we only have `Team.Authorize` before #17811, so maybe `Team.Authorize` is unnecessary now?
If it is unnecessary, maybe we need to do some changes to `GetUserOrgsPermissions` too.
https://github.com/go-gitea/gitea/blob/6706ac2a0f5f2fe4f8e2555be7e2a8b4d5946398/routers/api/v1/org/org.go#L105-L169

File: routers/api/packages/container/container.go [hunk 2]
```diff
@@ -143,6 +144,22 @@ func Authenticate(ctx *context.Context) {
...
+		}
+		if err := task.LoadJob(ctx); err != nil {
+			apiError(ctx, http.StatusInternalServerError, err)
+		}
+		if err := task.Job.LoadRun(ctx); err != nil {
+			apiError(ctx, http.StatusInternalServerError, err)
+		}
+		if err := task.Job.Run.LoadTriggerUser(ctx); err != nil {
+			apiError(ctx, http.StatusInternalServerError, err)
+		}
```
Comment (wolfogre) [inline routers/api/packages/container/container.go]: Is that correct?
If there's an error and the task is nil, and `task.Job` will panic, right?

**Review by wolfogre — APPROVED** [review]

**Comment by KN4CK3R** [comment]
This change needs to work for all package types not just the container registry.

File: routers/api/packages/api.go [hunk 3]
```diff
@@ -71,6 +72,10 @@ func CommonRoutes(ctx gocontext.Context) *web.Route {
 			ctx.Error(http.StatusUnauthorized, "authGroup.Verify")
 			return
 		}
+		if err := checkActionUser(ctx); err != nil {
+			ctx.Error(http.StatusInternalServerError, "checkActionUser")
```
Comment (lunny) [inline routers/api/packages/api.go]: `err` doesn't write to log?

**Review by wolfogre — CHANGES_REQUESTED** [review]
Sorry, I changed my idea. I ~don't like~ disagree:

`If user is ActionsUser, it will be changed to Task.Job.Run.TriggerUser to check the permission.`

Maybe it's convenient for users, but it's confusing and I don't see GitHub does in this way. 

**It may also lead to permission leakage. Think about this, if I am an administrator of gitea.com, and I help with a project hosted on gitea.com (just help, but I don't know/trust the owner). And my PR triggers a workflow, so the trigger is me, and the operations in it will get my administrator permissions!**

I hope we could make it clear:

- If an actions job did something, the doer is the actions bot user.
- If the operation need permissions, we grant the permissions to the actions bot user (could be done in the future).
- If it is inconvenient for you, it's recommended to use access token via secrets.

**Comment by yp05327** [comment]
I agree to add permission check to Action Bot User(maybe we can add an Internal User Type later?)
I will revert the code later.

**Review by yardenshoham — CHANGES_REQUESTED** [review]
dismiss my review

**Comment by yp05327** [comment]
You are right.
But if we use ActionUser to check permission, contributors can also get some special permissions of the Action Bot User which they don’t have.
This issue can be easily used to do some malicious things, as we can not control who create PRs in sometimes, so all contributors code should be untrusted.
And I think if the admin user want to contribute, they should at least know what the code is doing for, so the risk in this case is less.

**Comment by wolfogre** [comment]
> But if we use ActionUser to check permission, contributors can also get some special permissions of the Action Bot User which they don’t have.

https://github.com/go-gitea/gitea/blob/04347eb810689db799003cc342bbbc756716ff12/routers/web/repo/http.go#L210-L220

> This issue can be easily used to do some malicious things, as we can not control who create PRs in sometimes, so all codes from the contributors should be untrusted. It is an unpredictable risk.

https://github.com/go-gitea/gitea/pull/22803

> And I think if the admin user want to contribute, they should at least know what the code is doing for, so the risk in this case is more less. IMO normally admin user is used to manage the site, and not for contributing codes.

It's just a example. What if I am the owner of project A, and contribute to project B? So the workflows in project B will get my permission to project A.

> Maybe we need this:
[docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/enabling-features-for-your-repository/managing-github-actions-settings-for-a-repository#enabling-workflows-for-forks-of-private-repositories](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/enabling-features-for-your-repository/managing-github-actions-settings-for-a-repository#enabling-workflows-for-forks-of-private-repositories)

Yes, see https://github.com/go-gitea/gitea/pull/21937#discussion_r1090229868.

So before this, we need to give permission conservatively. Maybe it is a good choice to keep actions bot user without any permission to packages.

**Comment by sorenisanerd** [comment]
This feels scary. I don't think the tokens provided to actions should have the same permissions as any user, regardless of whether they're site admins, project owners, or random bypasses. We should always follow the principle of least privilege.

Each task should get a unique token with permissions as defined in the workflow file or fall back to reasonable defaults.

Github's defaults: https://docs.github.com/en/actions/security-guides/automatic-token-authentication#permissions-for-the-github_token
Documentation for adjusting the token privileges in actions: https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#permissions

**Comment by yp05327** [comment]
> So before this, we need to give permission conservatively. Maybe it is a good choice to keep actions bot user without any permission to packages.

I agree.  
So is it okey to only allow action bot user pull images from public registry in this PR, and add TODO for the permission check of private registry?

**Comment by wolfogre** [comment]
> > So before this, we need to give permission conservatively. Maybe it is a good choice to keep actions bot user without any permission to packages.
> 
> I agree.  
> So is it okey to only allow action bot user pull images from public registry in this PR, and add TODO for the permission check of private registry?

👍🏻I believe this is the optimal solution at the moment

**Comment by yp05327** [comment]
Removed permission check in this PR, and added `TODO` in https://github.com/go-gitea/gitea/pull/23879/files#diff-af878eff8f52037e8272d864e1b4d7ace9935f2149f3ed60fcf49cb6be891355R95

**Comment by wolfogre** [comment]
> Removed permission check in this PR, and added `TODO` in [#23879 (files)](https://github.com/go-gitea/gitea/pull/23879/files#diff-af878eff8f52037e8272d864e1b4d7ace9935f2149f3ed60fcf49cb6be891355R95)

Please update the description of this PR. 😄

**Comment by yp05327** [comment]
> > Removed permission check in this PR, and added `TODO` in [#23879 (files)](https://github.com/go-gitea/gitea/pull/23879/files#diff-af878eff8f52037e8272d864e1b4d7ace9935f2149f3ed60fcf49cb6be891355R95)
> 
> Please update the description of this PR. 😄

Done.

Maybe it is better to create a summary for this problem? It seems that the discussion is in many different issues/PRs comments.

**Review by wolfogre — APPROVED** [review]

**Review by lunny — APPROVED** [review]

**Review by KN4CK3R — APPROVED** [review]

**Pull request merged**